src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php

```diff
@@ -318,28 +330,43 @@ protected function instantiateObject(array &$data, $class, array &$context, \Ref
 
                         $params = array_merge($params, $data[$paramName]);
                     }
-                } elseif ($allowed && !$ignored && (isset($data[$key]) || array_key_exists($key, $data))) {
-                    $params[] = $data[$key];
-                    // don't run set for a parameter passed to the constructor
+
+                    continue;
+                }
+
+                if ($allowed && !$ignored && (isset($data[$key]) || array_key_exists($key, $data))) {
+                    $parameterData = $data[$key];
+                    if (null !== $constructorParameter->getClass()) {
+                        $parameterData = $this->serializer->deserialize($parameterData, $constructorParameter->getClass()->getName(), $format, $context);
+                    }
+
+                    // Don't run set for a parameter passed to the constructor
+                    $params[] = $parameterData;
                     unset($data[$key]);
-                } elseif ($constructorParameter->isDefaultValueAvailable()) {
+
+                    continue;
+                }
+
+                if ($constructorParameter->isDefaultValueAvailable()) {
                     $params[] = $constructorParameter->getDefaultValue();
-                } else {
-                    throw new RuntimeException(
-                        sprintf(
-                            'Cannot create an instance of %s from serialized data because its constructor requires parameter "%s" to be present.',
-                            $class,
-                            $constructorParameter->name
-                        )
-                    );
+
+                    continue;
                 }
+
+                throw new RuntimeException(
+                    sprintf(
+                        'Cannot create an instance of %s from serialized data because its constructor requires parameter "%s" to be present.',
+                        $class,
+                        $constructorParameter->name
+                    )
+                );
             }
 
             if ($constructor->isConstructor()) {
                 return $reflectionClass->newInstanceArgs($params);
-            } else {
```

Comment:
Yeah actually I think I'll revert everything for that part, it's just CS and although I think it makes a big difference it terms of readability, I fear that it will cause lots of conflicts later on. One solution would be do go back in earlier versions and fix that version after version, but it's rather tedious and slow.